src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -71,45 +71,87 @@ internal static class DotNetSdkLocationHelper
         }
 
         public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirectory, bool allowQueryAllRuntimes)
-        {            
-            foreach (var basePath in GetDotNetBasePaths(workingDirectory))
+        {
+            string? bestSdkPath;
+            string[] allAvailableSdks;
+            try
+            {
+                AddUnmanagedDllResolver();
+
+                bestSdkPath = GetSdkFromGlobalSettings(workingDirectory);
+                allAvailableSdks = GetAllAvailableSDKs();
+            }
+            finally
+            {
+                RemoveUnmanagedDllResolver();
+            }
+
+            Dictionary<Version, VisualStudioInstance?> versionInstanceMap = new();
+            foreach (var basePath in allAvailableSdks)
             {
                 var dotnetSdk = GetInstance(basePath, allowQueryAllRuntimes);
                 if (dotnetSdk != null)
                 {
-                    yield return dotnetSdk;
+                    // We want to return the best SDK first
+                    if (dotnetSdk.VisualStudioRootPath == bestSdkPath)
+                    {
+                        // We will add a null entry to the map to ensure we do not add the same SDK from a different location.
+                        versionInstanceMap[dotnetSdk.Version] = null;
+                        yield return dotnetSdk;
+                    }
+
+                    // Only add an SDK once, even if it's installed in multiple locations.
+                    if (!versionInstanceMap.ContainsKey(dotnetSdk.Version))
+                    {
+                        versionInstanceMap.Add(dotnetSdk.Version, dotnetSdk);
+                    }
                 }
             }
-        }
 
-        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
-        {
-            try
+            // We want to return the newest SDKs first. Using OfType will remove the null entry added if we found the best SDK.
+            var instances = versionInstanceMap.Values.OfType<VisualStudioInstance>().OrderByDescending(i => i.Version);
+            foreach (var instance in instances)
             {
-                AddUnmanagedDllResolver();
+                yield return instance;
+            }
 
-                string? bestSDK = GetSdkFromGlobalSettings(workingDirectory);
-                if (!string.IsNullOrEmpty(bestSDK))
+            // Returns the list of all available SDKs ordered by ascending version.
+            static string[] GetAllAvailableSDKs()
```

Comment:
This method and `GetSdkFromGlobalSettings` were only referenced from the inlined `GetDotNetBasePaths` so I made them local functions.